execution/subquery/subquery.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package subquery
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/thanos-io/promql-engine/execution/function"
+	"github.com/thanos-io/promql-engine/execution/model"
+	"github.com/thanos-io/promql-engine/parser"
+	"github.com/thanos-io/promql-engine/query"
+)
+
+// TODO: only instant subqueries for now.
+type subqueryOperator struct {
+	next        model.VectorOperator
+	pool        *model.VectorPool
+	funcExpr    *parser.Call
+	mint        int64
+	maxt        int64
+	currentStep int64
+
+	onceSeries sync.Once
+	series     []labels.Labels
+	acc        []float64
+}
+
+func NewSubqueryOperator(pool *model.VectorPool, next model.VectorOperator, opts *query.Options, funcExpr *parser.Call) model.VectorOperator {
+	return &subqueryOperator{
+		next:        next,
+		pool:        pool,
+		funcExpr:    funcExpr,
+		mint:        opts.Start.UnixMilli(),
+		maxt:        opts.End.UnixMilli(),
+		currentStep: opts.Start.UnixMilli(),
+	}
+}
+
+func (o *subqueryOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*subqueryOperator] %v()", o.funcExpr.Func.Name), []model.VectorOperator{o.next}
+}
+
+func (o *subqueryOperator) GetPool() *model.VectorPool { return o.pool }
+
+func (o *subqueryOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+	if err := o.initSeries(ctx); err != nil {
+		return nil, err
+	}
+
+ACC:
+	for {
```

Comment:
What is the reasoning behind this loop?

Reply:
the subquery might entail multiple steps of the underlying operator ( and since this PR so far only implements instant query subqueries for sum_over_time, its probably fine to just comsume all of them at the same time )

Reply:
I think it works like this for instant queries; range queries will be a bit more work though; wdyt of leaving it like this for instant queries with a big disclaimer comment and getting to work for range queries in subsequent PR?